### ISSUE TITLE
Version 2.0.0 (master) does not build with sleef / no_std features enabled

### ISSUE THREAD

**Issue by verpeteren** [arduano/simdeez#63]
@arduano did an overhaul of the SIMD traits. That was a big undertaking and has many nice improvements like operator overloading.
I am trying to port my stuff over to use the v2.0.0-dev3 (current master) branch and I noticed that there 3 problems:

1.  it does not build with no_std (`cargo build --features "no_std"`)
2.  it does not build with no_std (`cargo build --features "sleef"`)
3. it seems that only `scalar` is exposed. A simple test program with `sse2`, `sse41` or `avx2` will fi lead to `failed to resolve: could not find 'avx2`'in 'simdeez'`  ... `^^^^^ could not find 'avx2' in 'simdeez'`.

Please advice on how to assist with this.

**Comment by arduano** [comment]
Hi, You mention some valid concerns, along with some things that were intentional design decisions.

1. Not compiling with `no_std` is an oversight, I will attempt to fix it soon.
2. I haven't gotten around to porting sleef, it is a fairly weird library to include, but I know it needs to be done to maintain feature parity with the previous simdeez versions.
3. That was an intentional design decision. One of the main complaints I see for simdeez is that it's all unsafe, which is obviously caused by the fact that simdeez has no way of knowing whether individual operations are valid or not at runtime. My approach was to hide away any simd instruction set specific calls, and just allow:
  - Scalar (as it is always supported)
  - An general function that always picks the best instruction set, guaranteeing that the calls will be safe
  - Using `simd_unsafe_generate_all` to generate a list of unsafe functions for each instruction set

I made the structs associated with the instruction sets private, as I can't make using them directly safe, but at the same time I want to keep the API as safe as possible (without needing unsafe blocks literally everywhere).

Does `simd_unsafe_generate_all` fix the problem for you?

**Comment by arduano** [comment]
Oh also, the README example is outdated, please look at the examples folder for a functioning example. The `sum` function there just does basic SIMD addition from one vector into another, although modern CPUs can optimize that at runtime via pipelining just fine, so I also added a function there called `simd_get_sum` which does SIMD-based string parsing to parse all the numbers in the lists of characters and get the sum of them. Naturally, all of this needs more documentation, I just paused for a bit. I will look at it again soon though (today or next few days).

**Comment by arduano** [comment]
Ok I'm just looking at the `no_std` compile error, would you be familiar with how to do `neon` detection in no_std? It seems like it's possible to do x86 feature detection with `no_std`, but ARM doesn't seem to support it.

**Comment by verpeteren** [comment]
related to the feature detection, `rust doc` mentions:

```[is_aarch64_feature_detected](https://github.com/arduano/simdeez/issues/macro.is_aarch64_feature_detected.html)AArch64	This macro tests, at runtime, whether an aarch64 feature is enabled on aarch64 platforms. Currently most features are only supported on linux-based platforms.
```

**Comment by arduano** [comment]
Sorry, did you sent the above reply correctly? The code block seems empty

**Comment by verpeteren** [comment]
@arduano , I missed your comment about the missing code block. 
I edited the codeblock now.

**Comment by arduano** [comment]
Yeah the main issue with sleef is that it's an external C library that's more difficult to integrate. I am open to PRs though, if you have the capacity

**Comment by yzsolt** [comment]
I took a look at SIMD feature detection and it seems that in theory the [std_detect](https://docs.rs/crate/std_detect) crate could be used in `no_std` environments - it's the same implementation that's behind `std::arch`. They state:

> We intend to make std_detect more flexible and configurable in this regard to better serve the needs of #[no_std] targets.

 But currently it's not there yet: it requires a nightly rustc, a Git dependency (current 1.5.0 release doesn't build with recent rustc) and the `stdarch_internal` feature has to be enabled in simdeez, which is "strongly discouraged".

Looking at the issue from another perspective: does it really make sense to support runtime SIMD feature detection in `no_std` environments? I think `no_std` projects are well aware of their actual target architectures, so they can just use compile time code generation. So I would rather disable runtime detection on `no_std` by default, and _maybe_ make it available under a dedicated feature (implemented with std_detect).

If you agree, I'm happy to pursue this path and prepare a PR - and also take a look at the situation with `sleef`.

**Comment by arduano** [comment]
I believe no_std is mainly used in things like wasm, so it's probably ok to manually work around that (e.g. "no_std_wasm" feature flag)

Also there's zero reason to use nightly for simdeez, because simdeez purely exists to allow using simd features in stable. If you're using nightly, you might as well use the portable_simd feature, which does everything simdeez does but better
